File: src/main/java/com/chicman/view/signup/SignUp.java
package com.chicman.view.signup;

import com.chicman.ChicManUI;
import com.chicman.model.Customer;
import com.chicman.model.VerificationToken;
import com.chicman.repository.CustomerRepository;
import com.chicman.repository.VerificationTokenRepository;
import com.chicman.utility.Messages;
import com.chicman.utility.Utils;
import com.vaadin.icons.VaadinIcons;
import com.vaadin.server.Page;
import com.vaadin.server.UserError;
import com.vaadin.spring.annotation.SpringComponent;
import com.vaadin.spring.annotation.UIScope;
import com.vaadin.ui.*;

import java.util.List;
import java.util.UUID;

@SpringComponent
@UIScope
public class SignUp extends VerticalLayout {

	private final CustomerRepository customerRepo;
	private final VerificationTokenRepository verificationTokenRepo;
	private final ChicManUI eventHandler;
	private final TextField fullName;
	private final TextField email;
	private final PasswordField password;
    private final PasswordField passwordConfirm;
	private final Button btnSignUp;
    private Button btnUsernameCorrect;
    private Button btnEmailCorrect;
    private Button btnPasswordCorrect;
    private Button btnPasswordConfirmCorrect;
    private Button btnTemp;

    public SignUp(CustomerRepository customerRepo, VerificationTokenRepository verificationTokenRepo, ChicManUI eventHandler) {
		this.customerRepo = customerRepo;
		this.verificationTokenRepo = verificationTokenRepo;
		this.eventHandler = eventHandler;
		this.fullName = new TextField(Messages.get("auth.username"));
		this.email = new TextField(Messages.get("auth.email"));
		this.password = new PasswordField(Messages.get("auth.password"));
		this.passwordConfirm = new PasswordField(Messages.get("auth.password.confirm"));
		this.btnSignUp = new Button(Messages.get("auth.register"));
		this.btnUsernameCorrect = new Button();
        this.btnEmailCorrect = new Button();
        this.btnPasswordCorrect = new Button();
        this.btnPasswordConfirmCorrect = new Button();
        this.btnTemp = new Button();

		setupComponent();
		setEvent();
        JavaScript javaScript = Page.getCurrent().getJavaScript();
        javaScript.execute("enter2ClickButton(" + fullName.getId() + ", " + btnSignUp.getId() + ")");
        javaScript.execute("enter2ClickButton(" + email.getId() + ", " + btnSignUp.getId() + ")");
        javaScript.execute("enter2ClickButton(" + password.getId() + ", " + btnSignUp.getId() + ")");
        javaScript.execute("enter2ClickButton(" + passwordConfirm.getId() + ", " + btnSignUp.getId() + ")");
	}

	private void addUsernamePart() {
        HorizontalLayout usernamePart = new HorizontalLayout();
        usernamePart.setMargin(false);
        usernamePart.setSpacing(false);
        usernamePart.setWidth("100%");

        fullName.setId("fullNameSignUp");
        fullName.setWidth("100%");

        fullName.addFocusListener(e -> {
            fullName.setComponentError(null);
            if (fullName.getValue().length() > 0) {
                fullName.setCursorPosition(fullName.getValue().length());
            }
        });

        fullName.addBlurListener(e -> {
            if (fullName.getValue().length() > 0) {
                fullName.setComponentError(null);
                btnUsernameCorrect.setStyleName("validate-status-correct", true);
            } else {
                fullName.setComponentError(new UserError(Messages.get("auth.username.blank")));
                btnUsernameCorrect.removeStyleName("validate-status-correct");
            }
            Page.getCurrent().getJavaScript().execute("getFocused()");
        });

        usernamePart.addComponent(fullName);
        usernamePart.setComponentAlignment(fullName, Alignment.TOP_LEFT);
        usernamePart.setExpandRatio(fullName, 1);

        btnUsernameCorrect.setStyleName("borderless", true);
        btnUsernameCorrect.setStyleName("icon-only", true);
        btnUsernameCorrect.setStyleName("validate-status", true);
        btnUsernameCorrect.setIcon(VaadinIcons.CHECK);
        btnUsernameCorrect.setEnabled(false);
        btnUsernameCorrect.setVisible(true);

        usernamePart.addComponent(btnUsernameCorrect);
        usernamePart.setComponentAlignment(btnUsernameCorrect, Alignment.BOTTOM_RIGHT);

		addComponent(usernamePart);
		setComponentAlignment(usernamePart, Alignment.TOP_LEFT);
    }

    private void addEmailPart() {
        HorizontalLayout emailPart = new HorizontalLayout();
        emailPart.setMargin(false);
        emailPart.setSpacing(false);
        emailPart.setWidth("100%");

        email.setId("emailSignUp");
        email.setPlaceholder(Messages.get("auth.email.placeholder"));
        email.setWidth("100%");

        email.addFocusListener(e -> {
            email.setComponentError(null);
            if (email.getValue().length() > 0) {
                email.setCursorPosition(email.getValue().length());
            } else {
                email.setPlaceholder(Messages.get("auth.email.placeholder"));
            }
        });

        email.addBlurListener(e -> {
            if (Utils.isValidEmail(email)) {
                btnEmailCorrect.setStyleName("validate-status-correct", true);
            } else {
                btnEmailCorrect.removeStyleName("validate-status-correct");
            }

            Page.getCurrent().getJavaScript().execute("getFocused()");
        });

        emailPart.addComponent(email);
        emailPart.setComponentAlignment(email, Alignment.TOP_LEFT);
        emailPart.setExpandRatio(email, 1);

        btnEmailCorrect.setStyleName("borderless", true);
        btnEmailCorrect.setStyleName("icon-only", true);
        btnEmailCorrect.setStyleName("validate-status", true);
        btnEmailCorrect.setIcon(VaadinIcons.CHECK);
        btnEmailCorrect.setEnabled(false);
        btnEmailCorrect.setVisible(true);
        emailPart.addComponent(btnEmailCorrect);
        emailPart.setComponentAlignment(btnEmailCorrect, Alignment.BOTTOM_RIGHT);
		addComponent(emailPart);
		setComponentAlignment(emailPart, Alignment.TOP_LEFT);
    }

    private void addPasswordPart() {
        HorizontalLayout passwordPart = new HorizontalLayout();
        passwordPart.setMargin(false);
        passwordPart.setSpacing(false);
        passwordPart.setWidth("100%");

        password.setCaption(Messages.get("auth.password") +
                "&nbsp;" +
                "<span class='password-suggest'>" +
                Messages.get("auth.password.suggest") + "</span>");
        password.setCaptionAsHtml(true);

		password.setId("passwordSignUp");
		password.setWidth("100%");
		password.setMaxLength(10);

        password.addFocusListener(e -> {
            password.setComponentError(null);
            if (password.getValue().length() > 0) {
                password.setCursorPosition(password.getValue().length());
            }
        });

        password.addBlurListener(e -> {
            if (Utils.isValidPassword(password)) {
                btnPasswordCorrect.setStyleName("validate-status-correct", true);
            } else {
                btnPasswordCorrect.removeStyleName("validate-status-correct");
            }

            Page.getCurrent().getJavaScript().execute("getFocused()");
        });

        passwordPart.addComponent(password);
        passwordPart.setComponentAlignment(password, Alignment.TOP_LEFT);
        passwordPart.setExpandRatio(password, 1);

        btnPasswordCorrect.setStyleName("borderless");
        btnPasswordCorrect.setStyleName("icon-only", true);
        btnPasswordCorrect.setStyleName("validate-status", true);
        btnPasswordCorrect.setIcon(VaadinIcons.CHECK);
        btnPasswordCorrect.setEnabled(false);
        btnPasswordCorrect.setVisible(true);
        passwordPart.addComponent(btnPasswordCorrect);
        passwordPart.setComponentAlignment(btnPasswordCorrect, Alignment.BOTTOM_RIGHT);

		addComponent(passwordPart);
		setComponentAlignment(passwordPart, Alignment.TOP_LEFT);
    }

    private void addPasswordConfirmPart() {
        HorizontalLayout passwordConfirmPart = new HorizontalLayout();
        passwordConfirmPart.setMargin(false);
        passwordConfirmPart.setSpacing(false);
        passwordConfirmPart.setWidth("100%");

		passwordConfirm.setId("passwordConfirmSignUp");
		passwordConfirm.setWidth("100%");
		passwordConfirm.setMaxLength(10);

        passwordConfirm.addFocusListener(e -> {
            passwordConfirm.setComponentError(null);
            if (passwordConfirm.getValue().length() > 0) {
                passwordConfirm.setCursorPosition(passwordConfirm.getValue().length());
            }
        });

        passwordConfirm.addBlurListener(e -> {
            if (Utils.isValidPasswordConfirm(passwordConfirm, password)) {
                btnPasswordConfirmCorrect.addStyleName("validate-status-correct");
            } else {
                btnPasswordConfirmCorrect.removeStyleName("validate-status-correct");
            }

            Page.getCurrent().getJavaScript().execute("getFocused()");
        });

        passwordConfirmPart.addComponent(passwordConfirm);
        passwordConfirmPart.setComponentAlignment(passwordConfirm, Alignment.TOP_LEFT);
        passwordConfirmPart.setExpandRatio(passwordConfirm, 1);

        btnPasswordConfirmCorrect.setStyleName("borderless");
        btnPasswordConfirmCorrect.setStyleName("icon-only", true);
        btnPasswordConfirmCorrect.setStyleName("validate-status", true);
        btnPasswordConfirmCorrect.setIcon(VaadinIcons.CHECK);
        btnPasswordConfirmCorrect.setEnabled(false);
        btnPasswordConfirmCorrect.setVisible(true);
        passwordConfirmPart.addComponent(btnPasswordConfirmCorrect);
        passwordConfirmPart.setComponentAlignment(btnPasswordConfirmCorrect, Alignment.BOTTOM_RIGHT);

		addComponent(passwordConfirmPart);
		setComponentAlignment(passwordConfirmPart, Alignment.TOP_LEFT);
    }

    private void addBtnSignUpPart() {
        HorizontalLayout btnSignUpPart = new HorizontalLayout();
        btnSignUpPart.setMargin(false);
        btnSignUpPart.setSpacing(false);
        btnSignUpPart.setWidth("100%");

		btnSignUp.setStyleName("primary");
		btnSignUp.setWidth("100%");
		btnSignUp.setId("btnSignUp");
        btnSignUp.addBlurListener(e -> Page.getCurrent().getJavaScript().execute("getFocused()"));
		btnSignUpPart.addComponent(btnSignUp);
        btnSignUpPart.setComponentAlignment(btnSignUp, Alignment.TOP_LEFT);
        btnSignUpPart.setExpandRatio(btnSignUp, 1);

        btnTemp.setStyleName("borderless");
        btnTemp.setStyleName("icon-only", true);
        btnTemp.setStyleName("validate-status", true);
        btnTemp.setIcon(VaadinIcons.CHECK);
        btnTemp.setEnabled(false);
        btnTemp.setVisible(true);
        btnSignUpPart.addComponent(btnTemp);
        btnSignUpPart.setComponentAlignment(btnTemp, Alignment.BOTTOM_RIGHT);

        addComponent(btnSignUpPart);
        setComponentAlignment(btnSignUpPart, Alignment.TOP_LEFT);
    }
	
	private void setupComponent() {
	    addUsernamePart();
	    addEmailPart();
	    addPasswordPart();
	    addPasswordConfirmPart();
	    addBtnSignUpPart();

	    btnSignUp.focus();
	    fullName.setTabIndex(1);
	    email.setTabIndex(2);
	    password.setTabIndex(3);
	    passwordConfirm.setTabIndex(4);
	    btnSignUp.setTabIndex(5);
	}
	
	private void setEvent() {
		btnSignUp.addClickListener(e -> {
		    if (!Utils.isValidUsername(fullName)) return;
		    if (!Utils.isValidEmail(email)) return;
		    if (!Utils.isValidPassword(password)) return;
            if (!Utils.isValidPasswordConfirm(passwordConfirm, password)) return;

            List<Customer> storedCustomer = customerRepo.findByEmail(email.getValue());
            if (storedCustomer.size() > 0) {
                // email usable
                String error = Messages.get("error.sign_up.failed2") +
                        " " +
                        email.getValue() +
                        " " +
                        Messages.get("error.sign_up.failed3");

                Notification.show(Messages.get("error.sign_up.failed"), error, Notification.Type.HUMANIZED_MESSAGE);

                UserError userError = new UserError(Messages.get("error.email.already.used"));
                email.setComponentError(userError);
                btnEmailCorrect.removeStyleName("validate-status-correct");
            } else {
                Customer newUser = new Customer(fullName.getValue(), email.getValue(), password.getValue());
                customerRepo.save(newUser);

                String token = UUID.randomUUID().toString();
                VerificationToken verificationToken = new VerificationToken(token, newUser.getCustomerId());
                verificationTokenRepo.save(verificationToken);

                eventHandler.signedUp(verificationToken, newUser);
            }
		});
	}
}

File: src/main/java/com/chicman/view/signup/SignUp.kt
package com.chicman.view.signup

import com.chicman.ChicManUI
import com.chicman.bean.GuestBean
import com.chicman.repository.VerificationTokenRepository
import com.chicman.service.AuthService
import com.chicman.utility.LogUtils
import com.chicman.utility.Messages
import com.chicman.utility.Utils
import com.vaadin.icons.VaadinIcons
import com.vaadin.server.Page
import com.vaadin.server.UserError
import com.vaadin.spring.annotation.SpringComponent
import com.vaadin.spring.annotation.UIScope
import com.vaadin.ui.*

@SpringComponent
@UIScope
class SignUp(
    private val verificationTokenRepo: VerificationTokenRepository,
    private val eventHandler: ChicManUI,
    private val guestBean: GuestBean
) : VerticalLayout() {

    private val fullName: TextField = TextField(Messages.get("auth.username"))
    private val btnUsernameCorrect: Button = Button()

    private val email: TextField = TextField(Messages.get("auth.email"))
    private val btnEmailCorrect: Button = Button()

    private val password: PasswordField = PasswordField(Messages.get("auth.password"))
    private val btnPasswordCorrect: Button = Button()

    private val passwordConfirm: PasswordField = PasswordField(Messages.get("auth.password.confirm"))
    private val btnPasswordConfirmCorrect: Button = Button()

    private val btnSignUp: Button = Button(Messages.get("auth.register"))
    private val btnTempSignUp: Button = Button()


    init {
        setupComponent()
        setEvent()

        Page.getCurrent().javaScript.apply {
            execute("enter2ClickButton(${email.id}, ${btnSignUp.id})")
            execute("enter2ClickButton(${fullName.id}, ${btnSignUp.id})")
            execute("enter2ClickButton(${password.id}, ${btnSignUp.id})")
            execute("enter2ClickButton(${passwordConfirm.id}, ${btnSignUp.id})")
        }
    }

    private fun setupComponent() {
        addUsernamePart()
        addEmailPart()
        addPasswordPart()
        addPasswordConfirmPart()
        addBtnSignUpPart()

        btnSignUp.focus()
        fullName.tabIndex = 1
        email.tabIndex = 2
        password.tabIndex = 3
        passwordConfirm.tabIndex = 4
        btnSignUp.tabIndex = 5
    }

    private fun addUsernamePart() {
        val usernamePart = HorizontalLayout()
        usernamePart.setMargin(false)
        usernamePart.isSpacing = false
        usernamePart.setWidth("100%")
        fullName.id = "fullNameSignUp"
        fullName.setWidth("100%")
        fullName.addFocusListener {
            fullName.componentError = null
            if (fullName.value.isNotEmpty()) {
                fullName.cursorPosition = fullName.value.length
            }
        }
        fullName.addBlurListener {
            if (fullName.value.isNotEmpty()) {
                fullName.componentError = null
                btnUsernameCorrect.setStyleName("validate-status-correct", true)
            } else {
                fullName.componentError = UserError(Messages.get("auth.username.blank"))
                btnUsernameCorrect.removeStyleName("validate-status-correct")
            }
            Page.getCurrent().javaScript.execute("getFocused()")
        }
        usernamePart.addComponent(fullName)
        usernamePart.setComponentAlignment(fullName, Alignment.TOP_LEFT)
        usernamePart.setExpandRatio(fullName, 1f)
        btnUsernameCorrect.setStyleName("borderless", true)
        btnUsernameCorrect.setStyleName("icon-only", true)
        btnUsernameCorrect.setStyleName("validate-status", true)
        btnUsernameCorrect.icon = VaadinIcons.CHECK
        btnUsernameCorrect.isEnabled = false
        btnUsernameCorrect.isVisible = true
        usernamePart.addComponent(btnUsernameCorrect)
        usernamePart.setComponentAlignment(btnUsernameCorrect, Alignment.BOTTOM_RIGHT)
        addComponent(usernamePart)
        setComponentAlignment(usernamePart, Alignment.TOP_LEFT)
    }

    private fun addEmailPart() {
        val emailPart = HorizontalLayout()
        emailPart.setMargin(false)
        emailPart.isSpacing = false
        emailPart.setWidth("100%")
        email.id = "emailSignUp"
        email.placeholder = Messages.get("auth.email.placeholder")
        email.setWidth("100%")
        email.addFocusListener {
            email.componentError = null
            if (email.value.isNotEmpty()) {
                email.cursorPosition = email.value.length
            } else {
                email.placeholder = Messages.get("auth.email.placeholder")
            }
        }
        email.addBlurListener {
            if (Utils.isValidEmail(email)) {
                btnEmailCorrect.setStyleName("validate-status-correct", true)
            } else {
                btnEmailCorrect.removeStyleName("validate-status-correct")
            }
            Page.getCurrent().javaScript.execute("getFocused()")
        }
        emailPart.addComponent(email)
        emailPart.setComponentAlignment(email, Alignment.TOP_LEFT)
        emailPart.setExpandRatio(email, 1f)
        btnEmailCorrect.setStyleName("borderless", true)
        btnEmailCorrect.setStyleName("icon-only", true)
        btnEmailCorrect.setStyleName("validate-status", true)
        btnEmailCorrect.icon = VaadinIcons.CHECK
        btnEmailCorrect.isEnabled = false
        btnEmailCorrect.isVisible = true
        emailPart.addComponent(btnEmailCorrect)
        emailPart.setComponentAlignment(btnEmailCorrect, Alignment.BOTTOM_RIGHT)
        addComponent(emailPart)
        setComponentAlignment(emailPart, Alignment.TOP_LEFT)
    }

    private fun addPasswordPart() {
        val passwordPart = HorizontalLayout()
        passwordPart.setMargin(false)
        passwordPart.isSpacing = false
        passwordPart.setWidth("100%")
        password.caption = Messages.get("auth.password") +
            "&nbsp;" +
            "<span class='password-suggest'>" +
            Messages.get("auth.password.suggest") + "</span>"
        password.isCaptionAsHtml = true
        password.id = "passwordSignUp"
        password.setWidth("100%")
        password.maxLength = 10
        password.addFocusListener {
            password.componentError = null
            if (password.value.isNotEmpty()) {
                password.cursorPosition = password.value.length
            }
        }
        password.addBlurListener {
            if (Utils.isValidPassword(password)) {
                btnPasswordCorrect.setStyleName("validate-status-correct", true)
            } else {
                btnPasswordCorrect.removeStyleName("validate-status-correct")
            }
            Page.getCurrent().javaScript.execute("getFocused()")
        }
        passwordPart.addComponent(password)
        passwordPart.setComponentAlignment(password, Alignment.TOP_LEFT)
        passwordPart.setExpandRatio(password, 1f)
        btnPasswordCorrect.styleName = "borderless"
        btnPasswordCorrect.setStyleName("icon-only", true)
        btnPasswordCorrect.setStyleName("validate-status", true)
        btnPasswordCorrect.icon = VaadinIcons.CHECK
        btnPasswordCorrect.isEnabled = false
        btnPasswordCorrect.isVisible = true
        passwordPart.addComponent(btnPasswordCorrect)
        passwordPart.setComponentAlignment(btnPasswordCorrect, Alignment.BOTTOM_RIGHT)
        addComponent(passwordPart)
        setComponentAlignment(passwordPart, Alignment.TOP_LEFT)
    }

    private fun addPasswordConfirmPart() {
        val passwordConfirmPart = HorizontalLayout()
        passwordConfirmPart.setMargin(false)
        passwordConfirmPart.isSpacing = false
        passwordConfirmPart.setWidth("100%")
        passwordConfirm.id = "passwordConfirmSignUp"
        passwordConfirm.setWidth("100%")
        passwordConfirm.maxLength = 10
        passwordConfirm.addFocusListener {
            passwordConfirm.componentError = null
            if (passwordConfirm.value.isNotEmpty()) {
                passwordConfirm.cursorPosition = passwordConfirm.value.length
            }
        }
        passwordConfirm.addBlurListener {
            if (Utils.isValidPasswordConfirm(passwordConfirm, password)) {
                btnPasswordConfirmCorrect.addStyleName("validate-status-correct")
            } else {
                btnPasswordConfirmCorrect.removeStyleName("validate-status-correct")
            }
            Page.getCurrent().javaScript.execute("getFocused()")
        }
        passwordConfirmPart.addComponent(passwordConfirm)
        passwordConfirmPart.setComponentAlignment(passwordConfirm, Alignment.TOP_LEFT)
        passwordConfirmPart.setExpandRatio(passwordConfirm, 1f)
        btnPasswordConfirmCorrect.styleName = "borderless"
        btnPasswordConfirmCorrect.setStyleName("icon-only", true)
        btnPasswordConfirmCorrect.setStyleName("validate-status", true)
        btnPasswordConfirmCorrect.icon = VaadinIcons.CHECK
        btnPasswordConfirmCorrect.isEnabled = false
        btnPasswordConfirmCorrect.isVisible = true
        passwordConfirmPart.addComponent(btnPasswordConfirmCorrect)
        passwordConfirmPart.setComponentAlignment(btnPasswordConfirmCorrect, Alignment.BOTTOM_RIGHT)
        addComponent(passwordConfirmPart)
        setComponentAlignment(passwordConfirmPart, Alignment.TOP_LEFT)
    }

    private fun addBtnSignUpPart() {
        val btnSignUpPart = HorizontalLayout()
        btnSignUpPart.setMargin(false)
        btnSignUpPart.isSpacing = false
        btnSignUpPart.setWidth("100%")
        btnSignUp.styleName = "primary"
        btnSignUp.setWidth("100%")
        btnSignUp.id = "btnSignUp"
        btnSignUp.addBlurListener { Page.getCurrent().javaScript.execute("getFocused()") }
        btnSignUpPart.addComponent(btnSignUp)
        btnSignUpPart.setComponentAlignment(btnSignUp, Alignment.TOP_LEFT)
        btnSignUpPart.setExpandRatio(btnSignUp, 1f)
        btnTempSignUp.styleName = "borderless"
        btnTempSignUp.setStyleName("icon-only", true)
        btnTempSignUp.setStyleName("validate-status", true)
        btnTempSignUp.icon = VaadinIcons.CHECK
        btnTempSignUp.isEnabled = false
        btnTempSignUp.isVisible = true
        btnSignUpPart.addComponent(btnTempSignUp)
        btnSignUpPart.setComponentAlignment(btnTempSignUp, Alignment.BOTTOM_RIGHT)
        addComponent(btnSignUpPart)
        setComponentAlignment(btnSignUpPart, Alignment.TOP_LEFT)
    }

    private fun setEvent() {
        btnSignUp.addClickListener {
            if (!Utils.isValidUsername(fullName)) {
                LogUtils.error("invalid username")
                return@addClickListener
            }
            if (!Utils.isValidEmail(email)) {
                LogUtils.error("invalid email")
                return@addClickListener
            }
            if (!Utils.isValidPassword(password)) {
                LogUtils.error("invalid password")
                return@addClickListener
            }
            if (!Utils.isValidPasswordConfirm(passwordConfirm, password)) {
                LogUtils.error("confirm password not match")
                return@addClickListener
            }

            AuthService.register(
                guestBean.token!!,
                fullName.value,
                email.value,
                password.value,
                eventHandler.appPath ?: "?"
            )?.apply {
                LogUtils.error("register result: $redirectUrl")

//            if (1 > 0) {
//                // email usable
//                val error = Messages.get("error.sign_up.failed2") +
//                    " " +
//                    email.value +
//                    " " +
//                    Messages.get("error.sign_up.failed3")
//                Notification.show(Messages.get("error.sign_up.failed"), error, Notification.Type.HUMANIZED_MESSAGE)
//                val userError = UserError(Messages.get("error.email.already.used"))
//                email.componentError = userError
//                btnEmailCorrect.removeStyleName("validate-status-correct")
//            } else {
//                val newUser = Customer(fullName.value, email.value, password.value)
//                val token = UUID.randomUUID().toString()
//                val verificationToken = VerificationToken(token, newUser.customerId)
//                verificationTokenRepo.save(verificationToken)
//                eventHandler.signedUp(verificationToken, newUser)
//            }
            }
        }
    }

}
